evm/utils/test_builder.py

```diff
@@ -0,0 +1,309 @@
+from collections import OrderedDict
+from collections.abc import (
+    Iterable,
+    Mapping,
+)
+import copy
+import itertools
+import subprocess
+
+from cytoolz import (
+    assoc_in,
+    merge,
+)
+from eth_utils import (
+    decode_hex,
+    encode_hex,
+    force_text,
+    int_to_big_endian,
+    to_checksum_address,
+    apply_formatter_to_array,
+    combine_argument_formatters,
+    curried,
+)
+from evm.utils.fixture_tests import (
+    hash_log_entries,
+)
+
+from evm.db.state import (
+    MainAccountStateDB,
+)
+from evm.db.backends.memory import MemoryDB
+
+
+DEFAULT_ENV = OrderedDict([
+    ("currentCoinbase", decode_hex("0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba")),
+    ("currentDifficulty", 131072),
+    ("currentGasLimit", 1000000),
+    ("currentNumber", 1),
+    ("currentTimestamp", 1000),
+    ("previousHash", decode_hex(
+        "0x5e20a0453cecd065ea59c37ac63e079ee08998b6045136a8ce6635c7912ec0b6"
+    )),
+])
+
+ALL_FORKS = [
+    "Frontier",
+    "Homestead",
+    "EIP150",
+    "EIP158",
+    "Byzantium"
+]
+
+FILLED_WITH_TEMPLATE = "py-evm-{version}"
+
+
+def normalize_state(state):
+    if isinstance(state, Iterable) and not isinstance(state, Mapping):
+        state = state_list_to_dict(state)
+    for address, account_state_dict in state.items():
+        account_state_dict.setdefault("balance", 0)
+        account_state_dict.setdefault("nonce", 0)
+        account_state_dict.setdefault("code", b"")
+        account_state_dict.setdefault("storage", OrderedDict())
+    return state
+
+
+def state_list_to_dict(state_list):
+    d = {}
+    for state_item in state_list:
+        d = assoc_in(d, state_item[:-1], state_item[-1])
+    return d
+
+
+def merge_states(state1, state2):
+    new_state = copy.deepcopy(state2)
+    for key, value in state1.items():
+        if key not in new_state:
+            new_state[key] = copy.deepcopy(value)
+        elif isinstance(value, Mapping) and isinstance(new_state[key], Mapping):
+            new_state[key] = merge_states(new_state[key], value)
+    return new_state
+
+
+def without_duplicates(l):
+    return list(OrderedDict.fromkeys(l))
+
+
+def flatten(l):
+    return list(itertools.chain.from_iterable(l))
+
+
+def int_to_hex(i):
+    return encode_hex(int_to_big_endian(i))
+
+
+def get_version_from_git():
+    version = subprocess.check_output(["git", "describe"]).strip()
+    return force_text(version)
+
+
+env_formatter = curried.apply_formatters_to_dict({
```

Comment:
I love the formatters, they made the code much cleaner! The only one I missed was a dict key formatter (basically what `storage_formatter` and `state_formatter` do manually now), but I'm not sure if it's a good idea to add them to the library as it could destroy uniqueness of keys if one's not careful.

Reply:
You're looking for `eth_utils.apply_key_map`